src/testdrive/src/action/postgres/execute.rs

```diff
@@ -33,26 +33,32 @@ impl Action for ExecuteAction {
         Ok(())
     }
 
-    async fn redo(&self, _: &mut State) -> Result<(), String> {
-        let (client, conn) = tokio_postgres::connect(&self.connection, NoTls)
-            .await
-            .map_err(|e| format!("connecting to postgres: {}", e))?;
-        println!(
-            "Executing queries against PostgreSQL server at {}...",
-            self.connection
-        );
-        let conn_handle = tokio::spawn(conn);
+    async fn redo(&self, state: &mut State) -> Result<(), String> {
+        let client = if self.connection.starts_with("postgres://") {
+            postgres_client(&self.connection).await?
+        } else {
+            state
+                .postgres_clients
+                .remove(&self.connection)
+                .ok_or(format!("connection '{}' not found", &self.connection))?
+        };
+
         for query in &self.queries {
             println!(">> {}", query);
             client
                 .batch_execute(query)
                 .await
                 .map_err(|e| format!("executing postgres query: {}", e))?;
         }
-        drop(client);
-        conn_handle
-            .await
-            .unwrap()
-            .map_err(|e| format!("postgres connection error: {}", e))
+
+        if self.connection.starts_with("postgres://") {
+            drop(client);
```

Comment:
Now that client is a reference, this had to go. But I checked manually that connections are being closed as they go out of scope and are not leaking.